packages/java-edition/src/binder/index.ts

```diff
@@ -166,17 +169,38 @@ export function* getRels(
 	return undefined
 }
 
-export function dissectUri(uri: string, ctx: ContextBase) {
+export function dissectUri(uri: string, ctx: UriBinderContext) {
 	const regex =
 		/^data\/([^\/]+)\/((?:tags\/|worldgen\/)?[a-z_]+)\/(.*)(\.(?:mcfunction|json))$/
 
 	const rels = getRels(uri, ctx.roots)
+	const { customResources } = ctx.config.env
+
 	for (const rel of rels) {
 		const match = rel.match(regex)
 		if (!match) {
 			continue
 		}
-		const def = Categories.get(match[2])
+		let def = Categories.get(match[2])
+
+		if (!def && customResources.length > 0) {
+			// Try to match the category for custom resources.
+			const subFolders = match[3].split('/').slice(0, -1)
+
+			if (subFolders.length === 0) {
+				continue
+			} else {
+				for (let i = 0; i < subFolders.length - 1; i++) {
+					const customCategory = `${match[2]}/${
+						subFolders.slice(0, i + 1).join('/')
+					}`
+
+					def = Categories.get(customCategory)
+
+					// we don't break here because we want to find the most specific category
+				}
+			}
```

Comment:
we don't need the `else` block if the `if` is doing a `continue`. this lets us avoid unnecessary indentation

```suggestion
			if (subFolders.length === 0) {
				continue
			}
			
			for (let i = 0; i < subFolders.length - 1; i++) {
				const customCategory = `${match[2]}/${
					subFolders.slice(0, i + 1).join('/')
				}`

				def = Categories.get(customCategory)

				// we don't break here because we want to find the most specific category
			}
```

Reply:
actually, we dont need the explicit `subFolders.length === 0` check at all because the for-loop will be a no-op if it's an empty list